File: frontend/lib/models/coin.dart
/// Model representing an identified coin.
class Coin {
  final String id;
  final String name;
  final String country;
  final int? year;
  final String denomination;
  final double? faceValue;
  final String currency;
  final String? obverseDescription;
  final String? reverseDescription;
  final double confidence;

  const Coin({
    required this.id,
    required this.name,
    required this.country,
    this.year,
    required this.denomination,
    this.faceValue,
    required this.currency,
    this.obverseDescription,
    this.reverseDescription,
    required this.confidence,
  });

  factory Coin.fromJson(Map<String, dynamic> json) {
    return Coin(
      id: json['id'] as String,
      name: json['name'] as String,
      country: json['country'] as String,
      year: json['year'] as int?,
      denomination: json['denomination'] as String,
      faceValue: (json['face_value'] as num?)?.toDouble(),
      currency: json['currency'] as String,
      obverseDescription: json['obverse_description'] as String?,
      reverseDescription: json['reverse_description'] as String?,
      confidence: (json['confidence'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'country': country,
      'year': year,
      'denomination': denomination,
      'face_value': faceValue,
      'currency': currency,
      'obverse_description': obverseDescription,
      'reverse_description': reverseDescription,
      'confidence': confidence,
    };
  }

  /// Returns a confidence percentage string
  String get confidencePercent => '${(confidence * 100).toStringAsFixed(0)}%';

  /// Returns the year as a display string
  String get yearDisplay => year?.toString() ?? 'Unknown year';
}

/// Response from coin identification API
class CoinIdentificationResponse {
  final List<Coin> coins;
  final int totalCoinsDetected;
  final String modelUsed;

  const CoinIdentificationResponse({
    required this.coins,
    required this.totalCoinsDetected,
    required this.modelUsed,
  });

  factory CoinIdentificationResponse.fromJson(Map<String, dynamic> json) {
    return CoinIdentificationResponse(
      coins: (json['coins'] as List)
          .map((coin) => Coin.fromJson(coin as Map<String, dynamic>))
          .toList(),
      totalCoinsDetected: json['total_coins_detected'] as int,
      modelUsed: json['model_used'] as String,
    );
  }
}

File: frontend_react/src/types/coin.ts
/** A single identified coin from the backend. */
export interface Coin {
  id: string;
  name: string;
  country: string;
  year: number | null;
  denomination: string;
  face_value: number | null;
  currency: string;
  obverse_description: string | null;
  reverse_description: string | null;
  confidence: number;
  bbox?: number[] | null;
}

/** Response from the POST /api/v1/coins/identify endpoint. */
export interface CoinIdentificationResponse {
  coins: Coin[];
  total_coins_detected: number;
  model_used: string;
}

/** Response from GET / root endpoint. */
export interface AppInfo {
  name: string;
  version: string;
  description: string;
  docs: string;
  health: string;
}

/** Response from GET /api/v1/coins/health endpoint. */
export interface HealthResponse {
  status: string;
  model: string;
}

/** Data passed via React Router location state from Home to Results. */
export interface ResultsLocationState {
  coins: Coin[];
  totalCoinsDetected: number;
  modelUsed: string;
  imageUrl: string;
}
